File: Week7/단어변환.py
from collections import deque

def diff_one(a, b):
    # 두 단어가 정확히 한 글자만 다른지 체크
    cnt = 0
    for x, y in zip(a, b):
        if x != y:
            cnt += 1
    return cnt == 1

def solution(begin, target, words):
    # target이 words에 없으면 애초에 변환 불가
    if target not in words:
        return 0

    q = deque()
    q.append((begin, 0))  # (현재 단어, 지금까지 단계)
    visited = [False] * len(words)

    while q:
        cur, depth = q.popleft()

        # target 도착하면 그게 최소 단계
        if cur == target:
            return depth

        # 다음으로 갈 수 있는 단어들(한 글자만 다른 단어) 찾기
        for i, w in enumerate(words):
            if not visited[i] and diff_one(cur, w):
                visited[i] = True          # 큐에 넣을 때 방문 처리(중복 방지)
                q.append((w, depth + 1))   # 한 단계 증가

    return 0

File: Week7/단어변환.java
import java.util.*;

class Solution {

    // 두 단어가 한 글자만 다른지 확인
    public boolean diffOne(String a, String b) {
        int cnt = 0;

        for (int i = 0; i < a.length(); i++) {
            if (a.charAt(i) != b.charAt(i)) {
                cnt++;
            }
        }

        return cnt == 1;
    }

    public int solution(String begin, String target, String[] words) {

        // target이 words에 없으면 변환 불가
        boolean exists = false;
        for (String w : words) {
            if (w.equals(target)) {
                exists = true;
                break;
            }
        }

        if (!exists) return 0;

        // BFS 큐
        Queue<Node> q = new LinkedList<>();
        q.offer(new Node(begin, 0));

        boolean[] visited = new boolean[words.length];

        while (!q.isEmpty()) {

            Node current = q.poll();
            String curWord = current.word;
            int depth = current.depth;

            // target 도착
            if (curWord.equals(target)) {
                return depth;
            }

            // 다음 단어 탐색
            for (int i = 0; i < words.length; i++) {

                if (!visited[i] && diffOne(curWord, words[i])) {
                    visited[i] = true;
                    q.offer(new Node(words[i], depth + 1));
                }
            }
        }

        return 0;
    }
}

// 단어와 depth 저장 클래스
class Node {
    String word;
    int depth;

    Node(String word, int depth) {
        this.word = word;
        this.depth = depth;
    }
}
